qcfractal/storage_sockets/sqlalchemy_socket.py

```diff
@@ -1176,24 +1176,33 @@ def add_results(self, record_list: List[ResultRecord]):
 
                 if existing_res.get(idx) is None:
                     doc = ResultORM(**result.dict(exclude={"id"}))
+                    existing_res[idx] = doc
                     # results_list.append(doc)
                     results_list.append(doc)
                     new_record_idx.append(i)
                     # session.commit()  # TODO: faster if done in bulk
                     # result_ids.append("placeholder")
                     meta["n_inserted"] += 1
                 else:
-                    print("reached here")
-                    id_ = str(existing_res.get(idx))
-                    meta["duplicates"].append(id_)  # TODO
+                    doc = existing_res.get(idx)
+                    meta["duplicates"].append(doc)  # TODO
                     # If new or duplicate, add the id to the return list
-                    result_ids[i] = id_
+                    result_ids[i] = str(doc.id)
 
             session.add_all(results_list)
             session.commit()
-            for i, res in enumerate(results_list):
-                # print (type(new_records[i]))
-                result_ids[new_record_idx[i]] = str(res.id)
+            meta["duplicates"] = [str(doc.id) for doc in meta["duplicates"]] 
+            j, k = 0, 0
+            for i, _ in enumerate(result_ids):
```

Comment:
The previous method seemed much cleaner. Can you elaborate what you are trying to accomplish here?

Reply:
yeah unfortunately had to do some drastic changes, since when you add a new record, that new record should also be added to ```existing_res```, so a repetition of items in ```record_list``` be handled. Since all the new added items and the repeated ones in ```record_list``` do not have any id until ```session.add_all()```, we have to set their ```result_ids``` only after adding all, hence this block would exist.

Reply:
I mean the previous dictionary approach was much cleaner. Here you have a lot of offsets that could potentially go wrong depending on edge cases. Storing a list of `indices` and a dictionary of `indices : ids` will allow you to rebuild the `result_ids` in one pass without complex shifts. A list of duplicate `indices` could deliver similar ideas.

Reply:
I mean I thought of your approach and it is actually pretty reasonable and actually adds more to the code readability, but it requires you to use another for loop through the duplicate ```indices``` which might be degrading to the performance even though it should not be much. But then again, I guess you are right that the readability is worth the really small performance gain.

Reply:
Well, looping through a list of dict will take on the order of ~1 us while each query/check will take on the order of 1ms as shown by your timing test. Anything that doesn't hit the database is completely negligible in cost.